### PR TITLE
new iDeal payment method; request factory and unit tests

### PR DESCRIPTION
**Pull request by JudeNiroshan**
## Description

New request factory class built to support the iDeal payment method in payone. 
Unit tests are included. Integration tests are neglected as planned in the original plan.

### CODE REVIEW

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/IDealRequestFactory.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    @Nonnull
+    @Override
+    public BankTransferRequest createRequestInternal(@Nonnull RequestType requestType,
+                                                     @Nonnull PaymentWithCartLike paymentWithCartLike,
+                                                     @Nonnull TriFunction<RequestType, PayoneConfig, String, BankTransferRequest> requestConstructor) {
+        BankTransferRequest requestInternal = super.createRequestInternal(requestType, paymentWithCartLike, requestConstructor);
+        requestInternal.setBankGroupType(paymentWithCartLike.getPayment().getCustom().getFieldAsString(CustomFieldKeys.BANK_GROUP_TYPE));
+        return requestInternal;
+    }
+}
```
Comment (ahalberkamp) [inline service/src/main/java/com/commercetools/pspadapter/payone/mapping/IDealRequestFactory.java:26]: Do you need this Factory at all or can you just extend the method "createRequestInternal" of the superclass with the nullable parameter "BankGroupType"

**Review by ahalberkamp — CHANGES_REQUESTED** [review]

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/banktransfer/BankTransferRequest.java [hunk 2]
```diff
@@ -37,7 +39,7 @@
     public BankTransferRequest(@Nonnull final RequestType requestType,
                                @Nonnull final PayoneConfig config,
                                @Nonnull final String onlinebanktransfertype) {
-        super(config, requestType.getType(), ClearingType.PAYONE_PNT.getPayoneCode());
+        super(config, requestType.getType(), ClearingType.PAYONE_IDL.getPayoneCode());
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/banktransfer/BankTransferRequest.java:43]: I think this line should not be changed. As a co-incident, even. the enum has changed, the value is the same. I will revert back to `ClearingType.PAYONE_PNT`. The value is "sb" for bank-transfer requests.

**Review by lojzatran — CHANGES_REQUESTED** [review]
Did you at least manually test the whole process with the real running payone server? If so, can you give an ID of a payment that was used for testing, like I did in my PR? So we know that iDeal is really working?

**Comment by lojzatran** [comment]
In `CustomTypeBuilder.java` there's a type `PAYMENT_BANK_TRANSFER_IDEAL` which has fields `IBAN_FIELD` and `BIC_FIELD`. I don't think they will ever be necessary, so I think you can remove them. Maybe in the end you can use `PAYMENT_BANK_TRANSFER` type for IDEAL and don't have to create a new product type for that?

**Comment by JudeNiroshan** [comment]
@lojzatran So, shall we remove the custom Type definitions and reuse the `PAYMENT_BANK_TRANSFER` by adding the required fields as optional?

File: service/src/main/java/com/commercetools/pspadapter/tenant/TenantFactory.java [hunk 3]
```diff
@@ -291,6 +291,9 @@ protected PayoneRequestFactory createRequestFactory(@Nonnull PaymentMethod metho
             case BANK_TRANSFER_POSTFINANCE_EFINANCE:
                 return new PostFinanceBanktransferRequestFactory(tenantConfig);
 
+            case BANK_TRANSFER_IDEAL:
```
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply service/src/main/java/com/commercetools/pspadapter/tenant/TenantFactory.java:294]: e76b3aa

**Comment by lojzatran** [comment]
> @lojzatran So, shall we remove the custom Type definitions and reuse the `PAYMENT_BANK_TRANSFER` by adding the required fields as optional?

This is the question. WDYT?

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/IDealRequestFactory.java (hunk 1 above)
Comment (JudeNiroshan) [reply service/src/main/java/com/commercetools/pspadapter/payone/mapping/IDealRequestFactory.java]: 0e55dd951246895f73792445d3f19b524bc06d40

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/banktransfer/BankTransferRequest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply service/src/main/java/com/commercetools/pspadapter/payone/domain/payone/model/banktransfer/BankTransferRequest.java]: 7e6258452dda2ae95f1305625f72327fae9e2aad

**Review by ahalberkamp — APPROVED** [review]

File: service/src/main/java/com/commercetools/pspadapter/tenant/TenantFactory.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JudeNiroshan) [reply service/src/main/java/com/commercetools/pspadapter/tenant/TenantFactory.java:294]: Tested payment ID: `a6047df9-8e9e-4fac-9a0e-0d8cb4162938`

**Review by lojzatran — APPROVED** [review]

**Pull request merged**